### PR TITLE
🚨 Source Appsflyer: Migrate to V2 Appsflyer API

### LINKED ISSUE #37403: [source-appsflyer] host name issue
### Connector Name

source-appsflyer

### Connector Version

0.1.0

### What step the error happened?

During the sync

### Relevant information

while setting up the source as appsflyer i am receiving hostname error

### Relevant log output

```shell
2024-04-18 12:56:04 [32mINFO[m i.a.w.t.TemporalAttemptExecution(get):142 - Docker volume job log path: /tmp/workspace/633ce658-9d26-47fa-8505-01c0c1bf2fec/0/logs.log
2024-04-18 12:56:04 [32mINFO[m i.a.w.t.TemporalAttemptExecution(get):147 - Executing worker wrapper. Airbyte version: 0.42.1
2024-04-18 12:56:04 [32mINFO[m i.a.a.c.AirbyteApiClient(retryWithJitter):179 - Attempt 0 to save workflow id for cancellation
2024-04-18 12:56:04 [32mINFO[m i.a.c.i.LineGobbler(voidCall):149 - 
2024-04-18 12:56:04 [32mINFO[m i.a.c.EnvConfigs(getEnvOrDefault):1222 - Using default value for environment variable SIDECAR_KUBE_CPU_LIMIT: '2.0'
2024-04-18 12:56:04 [32mINFO[m i.a.c.EnvConfigs(getEnvOrDefault):1222 - Using default value for environment variable SOCAT_KUBE_CPU_LIMIT: '2.0'
2024-04-18 12:56:04 [32mINFO[m i.a.c.i.LineGobbler(voidCall):149 - ----- START CHECK -----
2024-04-18 12:56:04 [32mINFO[m i.a.c.i.LineGobbler(voidCall):149 - 
2024-04-18 12:56:04 [32mINFO[m i.a.c.EnvConfigs(getEnvOrDefault):1222 - Using default value for environment variable SIDECAR_KUBE_CPU_REQUEST: '0.1'
2024-04-18 12:56:04 [32mINFO[m i.a.c.EnvConfigs(getEnvOrDefault):1222 - Using default value for environment variable SOCAT_KUBE_CPU_REQUEST: '0.1'
2024-04-18 12:56:04 [32mINFO[m i.a.c.EnvConfigs(getEnvOrDefault):1222 - Using default value for environment variable LAUNCHDARKLY_KEY: ''
2024-04-18 12:56:04 [32mINFO[m i.a.c.EnvConfigs(getEnvOrDefault):1222 - Using default value for environment variable FEATURE_FLAG_CLIENT: ''
2024-04-18 12:56:04 [32mINFO[m i.a.c.i.LineGobbler(voidCall):149 - Checking if airbyte/source-appsflyer:0.1.0 exists...
2024-04-18 12:56:04 [32mINFO[m i.a.c.i.LineGobbler(voidCall):149 - airbyte/source-appsflyer:0.1.0 was found locally.
2024-04-18 12:56:04 [32mINFO[m i.a.w.p.DockerProcessFactory(create):130 - Creating docker container = source-appsflyer-check-633ce658-9d26-47fa-8505-01c0c1bf2fec-0-jkxlf with resources io.airbyte.config.ResourceRequirements@30222420[cpuRequest=,cpuLimit=,memoryRequest=,memoryLimit=] and allowedHosts null
2024-04-18 12:56:04 [32mINFO[m i.a.w.p.DockerProcessFactory(create):175 - Preparing command: docker run --rm --init -i -w /data/633ce658-9d26-47fa-8505-01c0c1bf2fec/0 --log-driver none --name source-appsflyer-check-633ce658-9d26-47fa-8505-01c0c1bf2fec-0-jkxlf --network host -v airbyte_workspace:/data -v /tmp/airbyte_local:/local -e DEPLOYMENT_MODE=OSS -e WORKER_CONNECTOR_IMAGE=airbyte/source-appsflyer:0.1.0 -e AUTO_DETECT_SCHEMA=true -e LAUNCHDARKLY_KEY= -e SOCAT_KUBE_CPU_REQUEST=0.1 -e SOCAT_KUBE_CPU_LIMIT=2.0 -e USE_STREAM_CAPABLE_STATE=true -e FIELD_SELECTION_WORKSPACES= -e AIRBYTE_ROLE= -e WORKER_ENVIRONMENT=DOCKER -e APPLY_FIELD_SELECTION=false -e WORKER_JOB_ATTEMPT=0 -e FEATURE_FLAG_CLIENT= -e AIRBYTE_VERSION=0.42.1 -e WORKER_JOB_ID=633ce658-9d26-47fa-8505-01c0c1bf2fec airbyte/source-appsflyer:0.1.0 check --config source_config.json
2024-04-18 12:56:04 [32mINFO[m i.a.w.i.VersionedAirbyteStreamFactory(create):110 - Reading messages from protocol version 0.2.0
2024-04-18 12:56:06 [1;31mERROR[m i.a.w.i.DefaultAirbyteStreamFactory(internalLog):190 - Check failed
2024-04-18 12:56:06 [32mINFO[m i.a.w.g.DefaultCheckConnectionWorker(run):120 - Check connection job received output: io.airbyte.config.StandardCheckConnectionOutput@39fef777[status=failed,message=ConnectionError(MaxRetryError("HTTPSConnectionPool(host='hq.appsflyer.com', port=443): Max retries exceeded with url: /export/id******86/partners_report/v5?api_token=********************KHqg&from=2024-04-18&to=2024-04-18&timezone=UTC (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f26fb1c3fd0>: Failed to establish a new connection: [Errno -2] Name does not resolve'))"))]
2024-04-18 12:56:06 [32mINFO[m i.a.w.t.TemporalAttemptExecution(get):169 - Stopping cancellation check scheduling...
2024-04-18 12:56:06 [32mINFO[m i.a.c.i.LineGobbler(voidCall):149 - 
2024-04-18 12:56:06 [32mINFO[m i.a.c.i.LineGobbler(voidCall):149 - ----- END CHECK -----
2024-04-18 12:56:06 [32mINFO[m i.a.c.i.LineGobbler(voidCall):149 -
```

### Contribute

- [ ] Yes, I want to contribute

### PR DESCRIPTION
**Pull request by anjay-goel**
## What

Migrate from obsolete V1 API to V2.
Fixes #37403, #35667, #32446.

## How

Appsflyer has deprecated & removed the V1 APIs. This PR migrates the connector to the API V2.
## User Impact

Users will have to replace the API V1 tokens with API V2 tokens as described [here](https://support.appsflyer.com/hc/en-us/articles/12399683708305-Bulletin-API-token-changes?query=token).
## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by natikgadzhi** [comment]
@AnjayGoel thank you for contributing! Running CI on this now.

I'd like to remove the dockerfile and setup.py while we're here — would you mind rebasing if/when I ping you?

**Comment by AnjayGoel** [comment]
> @AnjayGoel thank you for contributing! Running CI on this now.
> 
> I'd like to remove the dockerfile and setup.py while we're here — would you mind rebasing if/when I ping you?

Sure

**Comment by natikgadzhi** [comment]
Time to rebase! #38363 is in.

**Comment by AnjayGoel** [comment]
> Time to rebase! #38363 is in.

Done

**Comment by AnjayGoel** [comment]
> Is this connector running for all the streams? I think there are some schema/columns changes on the definition of the output. So maybe it would be nice to check and apply those changes. Here you have a reference about what I am talking about [aoelvp94@11d984a](https://github.com/aoelvp94/airbyte/commit/11d984ad817401f3581e0172b86177e4db1c7a06)


Hi @aoelvp94, thanks for pointing this out. It looks like some streams (`geo_report` & `partners_report`) have additional properties, based on the custom events the user sends to AppsFlyer (counter/unique users fields). So, we have to move to a dynamic schema for these streams. I will try to make those changes and update the PR soon

**Comment by AnjayGoel** [comment]
Hi @aoelvp94 
I've made the necessary changes to move to a dynamic schema. I've also tested all the streams, and everything seems to work fine.

File: airbyte-integrations/connectors/source-appsflyer/metadata.yaml
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-appsflyer/metadata.yaml:25]: We will need to bump this when everything else will be ready. Thank you for remembering to add the breaking change notice here!

**Review by natikgadzhi — COMMENTED** [review]
Looks good so far — I'll wait for other feedback, then we can apply formatters and see if we're ready to merge. 

One wild question: is AppsFlyer just a REST api under the hood? Should we yolo rewrite this to a low-code connector instead?

File: airbyte-integrations/connectors/source-appsflyer/metadata.yaml
Comment (AnjayGoel) [reply airbyte-integrations/connectors/source-appsflyer/metadata.yaml:25]: Sure, will bump this when the PR is ready for merging.

**Comment by AnjayGoel** [comment]
> Looks good so far — I'll wait for other feedback, then we can apply formatters and see if we're ready to merge.
> 
> One wild question: is AppsFlyer just a REST api under the hood? Should we yolo rewrite this to a low-code connector instead?

Yes, AppsFlyer is a REST API. If we can handle rate limiting & parse CSV responses then I think we should be able to rewrite this with the low-code connector instead.

**Comment by arielsnowik** [comment]
Hi. Which streams are going to be supported by this patch? Because I guess we are just fixing existing streams, not adding new ones by the updated API?

**Comment by arielsnowik** [comment]
Then I have another question. Can the dynamic schema infer data types from csv? Or it just finds the values. I don't understand why if that is the case why this code have a static schema in the first place

**Comment by AnjayGoel** [comment]
Hi @arielsnowik, The PR fixes the existing streams. I am not entirely sure if some new endpoints were introduced in v2. However, I have noticed that some endpoints, such as reinstalls, the pull API for organic source, and the ad-revenue data API, are missing in the current connector.

Also, The dynamic schema is only needed for the aggregate data APIs due to user-defined events present in the API response. It's not inferring the data types; it's just finding the additional fields and adding them to the schema as numerical type since the aggregate report only contains numbers.

**Comment by arielsnowik** [comment]
Thank you for the response.

We are analyzing using this patch (or to make our one) for a business project. Might be useful to understand which endpoints and which will not be supported (so we know what to develop if we need them), may we help with that? 

About the dynamic schema, so I understand we have a mix between static schema and dynamic schema for different fields, In our version of the patch we were trying to just make all static for simplicity of the solution, I don't know which would be better for long term support, because CSV does not support in a native way dynamic schema I believe (so making it dynamic sounds to me a little complicated I believe?) correct me if I am wrong in my reasoning.

Let me know what we can do to help you to make this patch complete/ well documented, as I said we are evaluating using/contributing to this patch or just making our own.

**Comment by AnjayGoel** [comment]
Hi @arielsnowik,
Following APIs are supported as of now:
* [Aggregated Data API]( https://dev.appsflyer.com/hc/reference/overview-11) (Except Reinstalls)
* [Raw Data API (Non Organic)](https://dev.appsflyer.com/hc/reference/get_app-id-installs-report-v5)
* [Raw Data(Retargeting)](https://dev.appsflyer.com/hc/reference/get_app-id-installs-retarget-v5)

Remaining ones are: 
* [Raw Data API (Organic)](https://dev.appsflyer.com/hc/reference/get_app-id-organic-installs-report-v5)
* [Ad Revenue](https://dev.appsflyer.com/hc/reference/get_app-id-ad-revenue-raw-v5)
* [Protect360](https://dev.appsflyer.com/hc/reference/get_app-id-blocked-installs-report-v5)

Adding support for them should be a simple task.

Also the issue with dynamic schema is because AppsFlyer is aggregating the In-App Events Statistics in the same report. The format of the aggregated report is like this:

```
Date, Country,Agency/PMD (af_prt),Media Source (pid),Campaign (c),Impressions,Clicks,CTR,Installs,Conversion Rate,Sessions,Loyal Users,Loyal Users/Installs,Total Revenue,Total Cost,ROI,ARPU,Average eCPI,in_app_event_1 (Unique users),in_app_event_1 (Event counter),in_app_event_1 (Sales in USD),in_app_event_2 (Unique users),in_app_event_2 (Event counter),in_app_event_2 (Sales in USD)...
```

Since everyone will have different In-App Events, its not possible to have a static schema definition. Also the events may be added/removed, changing the report. If we want to keep the schemas static, we can create a different stream for events statistics like this:
```
Date, Country,Agency/PMD (af_prt),Media Source (pid),Campaign (c), Event Name, Event counter,Sales in USD,Unique users
```

**Comment by arielsnowik** [comment]
Excellent. Thank you for the information about available methods. Would be nice to have that on readme.

I get it. In app events does not depend on the input, just the user, thus there the need dynamic to make thing generic. (Would be nice to have read that in appsflyer docs...)

However I believe also your idea for a static schema is nice, +1. Although maybe using another table with events so we don't waste space with repeated fields? Or a json field? Let me know what you think, and also what we can do to help.

Kind Regards

**Comment by AnjayGoel** [comment]
Hi @arielsnowik,
Thanks for the feedback. I will go ahead and add the additional stream, as discussed, soon. I will also document the currently supported endpoints.

**Comment by arielsnowik** [comment]
Hi. @AnjayGoel I have another question. When you refer to aggregated data api Reinstalls (not implemeted), I am not finding it either on appsflyer docs
image

May I be looking wrongly at docs?

**Comment by AnjayGoel** [comment]
> Hi. @AnjayGoel I have another question. When you refer to aggregated data api Reinstalls (not implemented), I am not finding it on appsflyer docs https://dev.appsflyer.com/hc/reference/overview-11 image
> 
> May I be looking wrong at docs? Or maybe it does not exist anymore?

My bad, Its in [Raw Data API](https://dev.appsflyer.com/hc/reference/get_app-id-reinstalls-v5)

**Comment by AnjayGoel** [comment]
Hi @arielsnowik, I have added the streams and updated the readme.

**Comment by AnjayGoel** [comment]
Hi @natikgadzhi , @aoelvp94,
Please let me know if anything else needs to be done before merging this PR.

**Comment by arielsnowik** [comment]
> Hi @arielsnowik, I have added the streams and updated the readme.

Nice, thank you!!

As long as the connector is properly tested and there is a clear way (by just reading readme...) for anyone to implement missing endpoints I think it would be ok to merge the PR. 
Let me know what do you (and the others) think.

Kind Regards

**Review by mauro-barletta — APPROVED** [review]
ok

**Comment by natikgadzhi** [comment]
Kicking off CI on this — I am not certain if we have test creds, but if the idea is that the current version is broken, than I just need formatters and static checks to pass.

**Comment by natikgadzhi** [comment]
@marcosmarxm, this one needs a breaking change migration guide:
```
source-appsflyer - ❌ Failed - Breaking changes must be accompanied by a migration guide: Migration guide file is missing for source-appsflyer. Please create a migration guide in ./docs/integrations/<connector-type>s/<connector-name>-migrations.md`.
source-appsflyer - ❌ Failed - Breaking change deadline should be a week in the future: The upgrade deadline for the breaking changes in 0.2.0 is less than 7 days from today. Please extend the deadline.
source-appsflyer - ❌ Failed - Connector version in metadata.yaml and pyproject.toml file must match: Version is 0.2.0 in metadata.yaml, but version is 0.1.1 in pyproject.toml. These two files have to be consistent.
Error: 3 checks failed
```

Can you link breaking change migration guide docs so @AnjayGoel can write one? 

@AnjayGoel, this also needs version to be set in both `pyproject.toml` and metadata.

**Comment by AnjayGoel** [comment]
@natikgadzhi I have made the changes

**Comment by marcosmarxm** [comment]
Triggered CI Tests.

**Review by marcosmarxm — APPROVED** [review]

**Comment by marcosmarxm** [comment]
Hello your feedback matters a lot to us. Can you spare just 3 minutes to complete a [survey](https://form.typeform.com/to/DOBqejte)? We're dedicated to making the contributor experience even better, and your input is key to achieving excellence. Thank you for helping us improve!

**Comment by aalkuatova** [comment]
Hello! Thank you so much for the recent update to the new version. I've been looking forward for this release, especially as I've encountered issues with the organic_in_app_events stream. After upgrading Appsflyer to version 0.2.0, unfortunately, I'm still experiencing the same problem with this stream.
The error I'm encountering in the logs is as follows:
`400 Client Error: Bad Request for URL: https://rawdata.appsflyer.com/export/token/lego_test__organic-in-app-events_2024-06-10_2024-06-11_JMQHXSDLCBLSWFW`
I've manually tested this endpoint and didn't encounter any issues with API limits or etc. Could you please help me in diagnosing and resolving this issue?
Thank you !

**Comment by marcosmarxm** [comment]
@aalkuatova do you have a github issue or can you opena. new one to further investigation?

**Comment by AnjayGoel** [comment]
Hi @marcosmarxm @aalkuatova , I have raised a PR to fix this: https://github.com/airbytehq/airbyte/pull/39407/

**Pull request merged**
